### PR TITLE
Handle Version Elements in Update

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
This enables the NuGetUpdater to handle "old-style" csproj files where the Version is a child of the PackageReference instead of just attributes.

Recommend reviewing with `&w=1` in the URL to disable whitespace diff.

### CODE REVIEW

File: nuget/helpers/lib/NuGetUpdater/.gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@
 artifacts/
 bin/
 obj/
+Properties/launchSettings.json
```
Comment (ryanbrandenburg) [inline nuget/helpers/lib/NuGetUpdater/.gitignore:5]: This is a QoL thing to prevent checking in launchSettings that get created when you're trying to run things locally.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/MSBuildHelperTests.cs [hunk 2]
```diff
@@ -187,6 +187,29 @@ public static IEnumerable<object[]> GetTopLevelPackageDependenyInfosTestData()
             }
         };
 
+        // version is a child-node of the package reference
+        yield return new object[]
```
Comment (ryanbrandenburg) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/MSBuildHelperTests.cs:191]: This already worked but I was already adding tests for this scenario so why not here too.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/SdkPackageUpdaterHelperTests.cs [hunk 3]
```diff
@@ -9,7 +9,7 @@ namespace NuGetUpdater.Core.Test.Utilities
     public class SdkPackageUpdaterHelperTests
     {
         [Fact]
-        public async void DirectoryBuildFilesAreOnlyPulledInFromParentDirectories()
+        public async Task DirectoryBuildFilesAreOnlyPulledInFromParentDirectories()
```
Comment (ryanbrandenburg) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/SdkPackageUpdaterHelperTests.cs:12]: Unrelated but async void always spooks me so I thought I'd change it while I was in the area.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/SdkPackageUpdaterTests.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        await SdkPackageUpdater.UpdateDependencyAsync(directory.DirectoryPath, projectPath, dependencyName, previousVersion, newDependencyVersion, isTransitive, logger);
+
+        // Assert
+        AssertContentsEqual(expectedContents, directory);
+    }
+
+    public static IEnumerable<object[]> GetDependencyUpdates()
+    {
+        // Simple case
+        yield return new object[]
```
Comment (ryanbrandenburg) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/SdkPackageUpdaterTests.cs:30]: I tried to make it simple to add additional cases we want to verify here but also didn't want to bog down the already complicated PR with a ton of examples.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs [hunk 5]
```diff
@@ -254,62 +254,133 @@ private static UpdateResult TryUpdateDependencyVersion(ImmutableArray<ProjectBui
 
         foreach (var buildFile in buildFiles)
         {
-            var updateAttributes = new List<XmlAttributeSyntax>();
-            var packageNodes = FindPackageNode(buildFile, dependencyName);
+            var updateNodes = new List<XmlNodeSyntax>();
```
Comment (ryanbrandenburg) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:257]: IMO my changes make this method a bit out of control WRT complexity, but when I took a swing at simplifying it it made the PR LESS reviewable instead of more.

I filed an issue on our side to make a separate pass at simplifying things here.

**Comment by jakecoffman** [comment]
I verified this fixes the issue we're seeing, is there someone on your team that needs to review it?

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/SdkPackageUpdaterHelperTests.cs (hunk 3 above)
Comment (abdulapopoola) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/SdkPackageUpdaterHelperTests.cs:12]: You bring back memories 😆 ; I'm with you on avoiding `async void`

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs [hunk 6]
```diff
@@ -254,62 +254,133 @@ private static UpdateResult TryUpdateDependencyVersion(ImmutableArray<ProjectBui
...
+            var packageNodes = FindPackageNodes(buildFile, dependencyName);
 
             var previousPackageVersion = previousDependencyVersion;
 
             foreach (var packageNode in packageNodes)
             {
                 var versionAttribute = packageNode.GetAttribute("Version", StringComparison.OrdinalIgnoreCase)
                     ?? packageNode.GetAttribute("VersionOverride", StringComparison.OrdinalIgnoreCase);
-                if (versionAttribute is null)
+                var versionElement = packageNode.Elements.FirstOrDefault(e => e.Name.Equals("Version", StringComparison.OrdinalIgnoreCase))
```
Comment (edgarrs) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:266]: my guess is that Elments cannot be null, right?

Comment (edgarrs) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:266]: And all elements have a non-null Name property, right?

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs [hunk 7]
```diff
@@ -254,62 +254,133 @@ private static UpdateResult TryUpdateDependencyVersion(ImmutableArray<ProjectBui
...
+                    {
+                        propertyNames.Add(versionAttribute.Value.Substring(2, versionAttribute.Value.Length - 3));
+                    }
+                    // Is this the case that the version is specified directly in the package node?
+                    else
+                    {
+                        var currentVersion = versionAttribute.Value.TrimStart('[', '(').TrimEnd(']', ')');
+                        if (currentVersion.Contains(',') || currentVersion.Contains('*'))
+                        {
+                            logger.Log($"    Found unsupported [{packageNode.Name}] version attribute value [{versionAttribute.Value}] in [{buildFile.RepoRelativePath}].");
```
Comment (jpinz) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:281]: q: can we not just use a `\t` here instead? It feels weird to just have a bunch of spaces at the beginning of the string.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs [hunk 8]
```diff
@@ -254,62 +254,133 @@ private static UpdateResult TryUpdateDependencyVersion(ImmutableArray<ProjectBui
...
                     }
-                    else if (currentVersion == newDependencyVersion)
-                    {
-                        logger.Log($"    Found correct [{packageNode.Name}] version attribute in [{buildFile.RepoRelativePath}].");
-                        foundCorrect = true;
-                    }
+                }
+                else
+                {
+                    // We weren't able to find the version node. Central package management?
```
Comment (jpinz) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:361]: log this scenario?

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs (hunk 6 above)
Comment (ryanbrandenburg) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:266]: The area is not null-annotated but my instinct was that they would just return an empty enumerable in that case, and looking at the code for XmlElementSyntax seems to support that hypothesis.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/SdkPackageUpdaterTests.cs (hunk 4 above)
Comment (jpinz) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/SdkPackageUpdaterTests.cs:30]: Personally, I don't mind extra stuff to review in a PR if it improves test coverage.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs [hunk 9]
```diff
@@ -254,62 +254,133 @@ private static UpdateResult TryUpdateDependencyVersion(ImmutableArray<ProjectBui
...
+                    // Is this the case that the version is specified directly in the package node?
+                    else
+                    {
+                        var currentVersion = versionAttribute.Value.TrimStart('[', '(').TrimEnd(']', ')');
+                        if (currentVersion.Contains(',') || currentVersion.Contains('*'))
+                        {
+                            logger.Log($"    Found unsupported [{packageNode.Name}] version attribute value [{versionAttribute.Value}] in [{buildFile.RepoRelativePath}].");
+                            foundUnsupported = true;
+                        }
+                        else if (currentVersion == previousDependencyVersion)
```
Comment (edgarrs) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:284]: Should we use an OrdinalComparison?

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs [hunk 10]
```diff
@@ -254,62 +254,133 @@ private static UpdateResult TryUpdateDependencyVersion(ImmutableArray<ProjectBui
...
 
-                // Is this the case where version is specified with property substitution?
-                if (versionAttribute.Value.StartsWith("$(") && versionAttribute.Value.EndsWith(")"))
-                {
-                    propertyNames.Add(versionAttribute.Value.Substring(2, versionAttribute.Value.Length - 3));
+                                logger.Log($"    Found incorrect peer [{packageNode.Name}] version attribute in [{buildFile.RepoRelativePath}].");
+                                updateNodes.Add(versionAttribute);
+                            }
+                        }
+                        else if (currentVersion == newDependencyVersion)
```
Comment (edgarrs) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:300]: Same, here should we use OrdinalComparison?

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs [hunk 11]
```diff
@@ -254,62 +254,133 @@ private static UpdateResult TryUpdateDependencyVersion(ImmutableArray<ProjectBui
...
+                    else
                     {
-                        var newVersion = NuGetVersion.Parse(newDependencyVersion);
-                        if (previousVersion < newVersion)
+                        var currentVersion = versionValue.TrimStart('[', '(').TrimEnd(']', ')');
+                        if (currentVersion.Contains(',') || currentVersion.Contains('*'))
+                        {
+                            logger.Log($"    Found unsupported [{packageNode.Name}] version node value [{versionValue}] in [{buildFile.RepoRelativePath}].");
+                            foundUnsupported = true;
+                        }
```
Comment (edgarrs) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:321]: I see why you mentioned this needs refactoring :)

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs [hunk 12]
```diff
@@ -254,62 +254,133 @@ private static UpdateResult TryUpdateDependencyVersion(ImmutableArray<ProjectBui
...
                         {
-                            previousPackageVersion = currentVersion;
+                            logger.Log($"    Found incorrect [{packageNode.Name}] version node in [{buildFile.RepoRelativePath}].");
+                            if (versionElement is XmlElementSyntax elementSyntax)
+                            {
+                                updateNodes.Add(elementSyntax);
+                            }
+                            else
+                            {
+                                throw new InvalidDataException("A concrete type was required for updateNodes. This should not happen.");
```
Comment (edgarrs) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:331]: Can you provide an example of when we could hit this exception?

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs (hunk 7 above)
Comment (ryanbrandenburg) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:281]: I agree (I'd actually go farther and say that log indentation/prefix should be managed by BeginScope or a similar concept) but right now I'm just doing what matches the rest of the file. If we wanted to revisit how the NuGetUpdater handles logging as a separate PR/set of PRs I'd be totally down with that.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs [hunk 13]
```diff
@@ -398,12 +469,18 @@ private static UpdateResult TryUpdateDependencyVersion(ImmutableArray<ProjectBui
                 : foundUnsupported
                     ? UpdateResult.NotSupported
                     : UpdateResult.NotFound;
+
+        static XmlElementSyntax WithConcreateContent(XmlElementSyntax element, string content)
```
Comment (edgarrs) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:473]: Why is this needed?

Comment (ryanbrandenburg) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:473]: Originally I was thinking it would get used multiple times, but since it only got used once I've folded it into it's single use location in my local changes (coming soon!)

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs (hunk 12 above)
Comment (ryanbrandenburg) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:331]: Adding a comment: `// This only exists for completeness in case we ever add a new type of node we don't want to silently ignore them.`

**Review by edgarrs — APPROVED** [review]

**Review by jpinz — APPROVED** [review]
:shipit:

**Comment by ryanbrandenburg** [comment]
@jakecoffman our review desires are satisfied. Feel free to merge/ship when you're ready.
